client/core/core.go

```diff
@@ -130,11 +130,158 @@ func (dc *dexConnection) findOrder(oid order.OrderID) (tracker *trackedTrade, pr
+	}
+	return nil
+}
+
+// trackerMatches are an intermediate structure used by the dexConnection to
+// sort incoming match notifications.
+type trackerMatches struct {
+	tracker    *trackedTrade
+	msgMatches []*msgjson.Match
+	cancel     *msgjson.Match
+}
+
+// parseMatches sorts the list of matches and associates them with a trade.
+func (dc *dexConnection) parseMatches(msgMatches []*msgjson.Match) (map[order.OrderID]*trackerMatches, []msgjson.Acknowledgement, error) {
+	var acks []msgjson.Acknowledgement
+	matches := make(map[order.OrderID]*trackerMatches)
+	var errs []string
+	for _, msgMatch := range msgMatches {
+		var oid order.OrderID
+		copy(oid[:], msgMatch.OrderID)
+		tracker, _, isCancel := dc.findOrder(oid)
+		if tracker == nil {
+			errs = append(errs, "order "+oid.String()+" not found")
+			continue
+		}
+		sigMsg, err := msgMatch.Serialize()
+		if err != nil {
+			errs = append(errs, err.Error())
+			continue
+		}
+		err = dc.acct.checkSig(sigMsg, msgMatch.Sig)
+		if err != nil {
+			return nil, nil, fmt.Errorf("parseMatches: %v", err)
+		}
+		sig, err := dc.acct.sign(sigMsg)
+		if err != nil {
+			errs = append(errs, err.Error())
+			continue
+		}
+		acks = append(acks, msgjson.Acknowledgement{
+			MatchID: msgMatch.MatchID,
+			Sig:     sig,
+		})
+
+		trackerID := tracker.ID()
+		match := matches[trackerID]
+		if match == nil {
+			match = &trackerMatches{
+				tracker: tracker,
+			}
+			matches[trackerID] = match
+		}
+		if isCancel {
+			err = tracker.processCancelMatch(msgMatch)
+		} else {
+			match.msgMatches = append(match.msgMatches, msgMatch)
+		}
+	}
+	var err error
+	if len(errs) > 0 {
+		err = fmt.Errorf("match errors: %s", strings.Join(errs, ", "))
+	}
+	return matches, acks, err
+}
+
+// runMatches runs the sorted matches returned from parseMatches.
+func (dc *dexConnection) runMatches(matches map[order.OrderID]*trackerMatches) error {
+	for _, match := range matches {
+		var err error
+		if len(match.msgMatches) > 0 {
+			err = match.tracker.negotiate(match.msgMatches)
+			if err == nil {
+				err = match.tracker.tick()
+			}
+		}
+		if err != nil {
+			return err
+		}
+		if match.cancel != nil {
+			err = match.tracker.processCancelMatch(match.cancel)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// tickAsset checks open matches related to a specific asset for needed action.
+func (dc *dexConnection) tickAsset(assetID uint32) {
+	dc.tradeMtx.RLock()
+	defer dc.tradeMtx.RUnlock()
+	for _, trade := range dc.trades {
+		if trade.Base() == assetID || trade.Quote() == assetID {
+			err := trade.tick()
+			if err != nil {
+				log.Errorf("%s tick error: %v", dc.acct.url, err)
+			}
+		}
+	}
+}
+
+// blockWaiter is a message waiting to be stamped, signed, and sent once a
+// specified coin has the requisite confirmations. The blockWaiter is similar to
+// dcrdex/server/blockWaiter.Waiter, but is different enough to warrant a
 // separate type.
-type coinWaiter struct {
+type blockWaiter struct {
```

Comment:
Planning to adapt the `coinwaiter.Waiter` for wider use (see `TODO:` note in `processAudit`). This was named `coinWaiter`, but was waiting for confirmations, not latency, hence the rename.